### PR TITLE
A first couple of PreDef-like methods with tests.

### PR DESCRIPTION
**Pull request by tfenne**

### CODE REVIEW

**Comment by tfenne** [comment]
Thinking out loud - perhaps these should be in package `dagr` and not `dagr.core` (but still within the core project).

File: core/src/main/scala/dagr/core/DagrDef.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  /** An exception that implies that code is unreachable. */
+  private class UnreachableException(message: String) extends IllegalStateException(message)
+
+  /**
+    * A terse way to throw an `UnreachableException` that can be used where any type is expected,
+    * e.g. `Option(thing) getOrElse unreachable("my thing is never null")`
+    *
+    * @param message an optional message
+    */
+  def unreachable(message: => String = ""): Nothing = throw new UnreachableException(message)
```
Comment (nh13) [inline core/src/main/scala/dagr/core/DagrDef.scala]: I confirmed this works on `getOrElse`, so this makes me really happy.

File: core/src/main/scala/dagr/core/DagrDef.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package dagr.core
+
+/**
+  * Object that is designed to be imported with `import DagrDef._` in any/all dagr classes
+  * much like the way that scala.PreDef is imported in all files automatically.
+  */
```
Comment (nh13) [inline core/src/main/scala/dagr/core/DagrDef.scala]: Perhaps a warning about not shoving things in here arbitrarily.

**Comment by nh13** [comment]
:+1:

**Pull request merged**